File: sorting-searching/searching-algorithms/interpolation-search/interpolation.py
from typing import List

def interpolation_search(arr: List[int], target: int) -> int:
    """Returns the index of target if found, else -1.
    PRECONDITION: arr must be sorted in ascending order."""
    low, high = 0, len(arr) - 1

    while low <= high and arr[low] <= target <= arr[high]:
        if low == high:
            return low if arr[low] == target else -1

        if arr[high] == arr[low]:
            return low if arr[low] == target else -1

        # Interpolation formula using integer floor division
        pos = low + ((target - arr[low]) * (high - low)) // (arr[high] - arr[low])

        if arr[pos] == target:
            return pos
        elif arr[pos] < target:
            low = pos + 1
        else:
            high = pos - 1

    return -1

if __name__ == "__main__":
    arr = [10, 12, 13, 16, 18, 19, 20, 21, 22, 23, 24, 33, 35, 42, 47]
    target = 18
    result = interpolation_search(arr, target)

    if result != -1:
        print(f"Element found at index: {result}")
    else:
        print("Element not found.")

File: sorting-searching/searching-algorithms/interpolation-search/interpolation.cpp
#include <iostream>
#include <vector>

// Returns index of target if found, else -1
// PRECONDITION: arr must be sorted in ascending order
int interpolationSearch(const std::vector<int>& arr, int target) {
    int low = 0;
    int high = static_cast<int>(arr.size()) - 1;

    while (low <= high && target >= arr[low] && target <= arr[high]) {
        // If only one element remains
        if (low == high) {
            return (arr[low] == target) ? low : -1;
        }

        // Prevent division by zero
        if (arr[high] == arr[low]) {
            return (arr[low] == target) ? low : -1;
        }

        // Interpolation formula
        int pos = low + static_cast<int>(
            ((double)(target - arr[low]) * (high - low)) / (arr[high] - arr[low])
        );

        if (arr[pos] == target) return pos;
        else if (arr[pos] < target) low = pos + 1;
        else high = pos - 1;
    }
    return -1;
}

int main() {
    std::vector<int> arr = {10, 12, 13, 16, 18, 19, 20, 21, 22, 23, 24, 33, 35, 42, 47};
    int target = 18;
    int result = interpolationSearch(arr, target);

    if (result != -1)
        std::cout << "Element found at index: " << result << "\n";
    else
        std::cout << "Element not found.\n";

    return 0;
}